cpp/src/arrow/array/builder_base.h

```diff
@@ -177,6 +177,9 @@ class ARROW_EXPORT ArrayBuilder {
   /// The given array must be the same type as the builder.
   virtual Status AppendArraySlice(const ArraySpan& array, int64_t offset,
                                   int64_t length) {
+    ARROW_UNUSED(array);
+    ARROW_UNUSED(offset);
+    ARROW_UNUSED(length);
```

Comment:
How about using `ARROW_ARG_UNUSED()` instead?

```suggestion
  virtual Status AppendArraySlice(const ArraySpan& ARROW_ARG_UNUSED(array), int64_t ARROW_ARG_UNUSED(offset),
                                  int64_t ARROW_ARG_UNUSED(length)) {
```

Reply:
Sure that works. Out of curiosity now that we are on C++17 any reason not to use `[[maybe_unused]]`?

Reply:
I think that we can use `[[maybe_unused]]`.
Could you open a new issue for this? Let's discuss this on the issue.

Reply:
Can you just try to use `[[maybe_unused]]` in this PR?